File: solutions/2026-08-05/javascript/2026-08-05-remove-methods-from-project.js
// Summary: Identify and remove a group of methods (including method k and its dependents) from a project if no external methods invoke any methods within this suspicious group.
// Link: https://leetcode.com/problems/remove-methods-from-project/
// Approach:
// 1. Represent the method invocations as a directed graph where an edge from `a` to `b` means `a` invokes `b`.
// 2. Identify all "suspicious" methods. A method is suspicious if it's method `k` or if it's invoked, directly or indirectly, by method `k`. We can find these using a Breadth-First Search (BFS) or Depth-First Search (DFS) starting from `k`.
// 3. To check the removal condition, we need to know which methods are *invoked by* other methods. This suggests building a reverse graph (or an adjacency list where each node stores who invokes it).
// 4. For each suspicious method, check if it's invoked by any method *not* in the suspicious set.
// 5. If *any* suspicious method is invoked by a non-suspicious method, then the entire suspicious set cannot be removed. In this case, return all methods.
// 6. Otherwise, if no suspicious method is invoked by a non-suspicious method, return the set of non-suspicious methods.
//
// Detailed Steps:
// a. Build the adjacency list for the invocation graph (`adj`) and the reverse graph (`reverseAdj`).
// b. Perform a BFS or DFS starting from `k` on the `adj` graph to find all suspicious methods and store them in a `suspiciousSet`.
// c. Iterate through all methods in the `suspiciousSet`. For each suspicious method `s`, check its incoming edges in `reverseAdj`.
// d. If `s` is invoked by any method `invoker` such that `invoker` is *not* in `suspiciousSet`, then we cannot remove the suspicious methods. Return all `n` methods.
// e. If the loop completes without finding any external invoker for a suspicious method, then all suspicious methods can be removed. Construct and return an array of non-suspicious methods.
//
// Time Complexity:
// Building graphs: O(N + E), where N is the number of methods and E is the number of invocations.
// Finding suspicious methods (BFS/DFS): O(N + E).
// Checking for external invokers: In the worst case, we might iterate through all edges in the reverse graph. However, we only do this for suspicious nodes and check their direct invokers. If a suspicious node has `d_in` incoming edges, we check `d_in` potential invokers. The total number of incoming edges for all suspicious nodes can be at most E. So, checking takes O(N + E) in total.
// Overall: O(N + E).
//
// Space Complexity:
// Adjacency lists: O(N + E).
// `suspiciousSet`: O(N).
// `visited` set for BFS/DFS: O(N).
// Overall: O(N + E).
var removeMethods = function(n, k, invocations) {
    // Adjacency list to represent method invocations: adj[a] contains methods invoked by a.
    const adj = Array(n).fill(0).map(() => []);
    // Reverse adjacency list to represent who invokes a method: reverseAdj[b] contains methods that invoke b.
    const reverseAdj = Array(n).fill(0).map(() => []);

    // Build the graphs
    for (const [u, v] of invocations) {
        adj[u].push(v);
        reverseAdj[v].push(u);
    }

    // Use a Set to store all suspicious methods for efficient lookups.
    const suspiciousSet = new Set();
    // Queue for BFS to find all methods suspicious directly or indirectly via k.
    const queue = [k];
    // Keep track of visited nodes during BFS to avoid cycles and redundant processing.
    const visited = new Set([k]);
    suspiciousSet.add(k); // Method k is always suspicious.

    // BFS to find all methods that are invoked by k (directly or indirectly).
    while (queue.length > 0) {
        const currentMethod = queue.shift();

        // Explore methods invoked by the current method.
        for (const neighbor of adj[currentMethod]) {
            if (!visited.has(neighbor)) {
                visited.add(neighbor);
                suspiciousSet.add(neighbor);
                queue.push(neighbor);
            }
        }
    }

    // Check if any suspicious method is invoked by a non-suspicious method.
    let canRemove = true;
    // Iterate through each method that is potentially suspicious.
    for (const suspiciousMethod of suspiciousSet) {
        // Check all methods that directly invoke the current suspicious method.
        for (const invoker of reverseAdj[suspiciousMethod]) {
            // If an invoker is NOT in the suspicious set, then the suspicious group cannot be removed.
            if (!suspiciousSet.has(invoker)) {
                canRemove = false;
                break; // No need to check further for this suspicious method.
            }
        }
        if (!canRemove) {
            break; // No need to check further suspicious methods.
        }
    }

    // If suspicious methods can be removed, return the remaining non-suspicious methods.
    if (canRemove) {
        const remainingMethods = [];
        // Iterate through all methods from 0 to n-1.
        for (let i = 0; i < n; i++) {
            // If a method is not in the suspicious set, it remains.
            if (!suspiciousSet.has(i)) {
                remainingMethods.push(i);
            }
        }
        return remainingMethods;
    } else {
        // If suspicious methods cannot be removed, return all methods.
        const allMethods = [];
        for (let i = 0; i < n; i++) {
            allMethods.push(i);
        }
        return allMethods;
    }
};
```

File: solutions/2026-08-05/cpp/2026-08-05-remove-methods-from-project.cpp
// Problem: Remove Methods From Project
// Link: https://leetcode.com/problems/remove-methods-from-project/
// Approach:
// 1. Identify all "suspicious" methods. A method is suspicious if it's `k` or if it's invoked (directly or indirectly) by `k`. This can be found using a graph traversal (DFS or BFS) starting from `k`.
// 2. Build an "invoker" graph to determine which methods invoke other methods. This is the reverse of the `invocations` graph.
// 3. For each suspicious method, check if any method *outside* the set of suspicious methods invokes it.
// 4. If *any* suspicious method is invoked by a non-suspicious method, then no suspicious methods can be removed, and we return all methods.
// 5. Otherwise, if all suspicious methods are only invoked by other suspicious methods (or not at all by non-suspicious methods), then all suspicious methods can be removed. We return the set of non-suspicious methods.
//
// Time Complexity:
// - Building adjacency lists for both invocation and invoker graphs: O(N + E), where N is the number of methods and E is the number of invocations.
// - DFS/BFS to find all methods reachable from `k`: O(N + E).
// - Iterating through suspicious methods and checking their invokers: In the worst case, each suspicious method might have multiple invokers. Checking if an invoker is suspicious takes O(1) if using a set. Total: O(N + E).
// Overall Time Complexity: O(N + E)
//
// Space Complexity:
// - Adjacency lists for invocation and invoker graphs: O(N + E).
// - Set to store suspicious methods: O(N).
// - Visited array/set for DFS/BFS: O(N).
// - Result vector: O(N).
// Overall Space Complexity: O(N + E)

#include <vector>
#include <queue>
#include <unordered_set>
#include <algorithm>

class Solution {
public:
    std::vector<int> kthSmallest(int n, int k, std::vector<std::vector<int>>& invocations) {
        // Adjacency list to represent method invocations: adj[u] contains methods invoked by u
        std::vector<std::vector<int>> adj(n);
        // Adjacency list to represent inverse invocations (who invokes a method): rev_adj[v] contains methods that invoke v
        std::vector<std::vector<int>> rev_adj(n);

        // Build the adjacency lists
        for (const auto& inv : invocations) {
            int u = inv[0]; // invoking method
            int v = inv[1]; // invoked method
            adj[u].push_back(v);
            rev_adj[v].push_back(u);
        }

        // Step 1: Find all suspicious methods using BFS starting from k
        std::unordered_set<int> suspicious_methods;
        std::queue<int> q;
        std::vector<bool> visited_suspicious(n, false);

        // Start BFS from method k
        q.push(k);
        visited_suspicious[k] = true;
        suspicious_methods.insert(k);

        while (!q.empty()) {
            int current_method = q.front();
            q.pop();

            // Explore methods invoked by current_method
            for (int neighbor : adj[current_method]) {
                if (!visited_suspicious[neighbor]) {
                    visited_suspicious[neighbor] = true;
                    q.push(neighbor);
                    suspicious_methods.insert(neighbor);
                }
            }
        }

        // Step 2 & 3: Check if any suspicious method is invoked by a non-suspicious method
        bool can_remove_suspicious = true;
        for (int suspicious_method : suspicious_methods) {
            // For each suspicious method, check its direct invokers
            for (int invoker : rev_adj[suspicious_method]) {
                // If an invoker is NOT in the set of suspicious methods,
                // then this suspicious method cannot be removed.
                if (suspicious_methods.find(invoker) == suspicious_methods.end()) {
                    can_remove_suspicious = false;
                    break; // No need to check further invokers for this suspicious_method
                }
            }
            if (!can_remove_suspicious) {
                break; // If one suspicious method can't be removed, then none can be removed.
            }
        }

        // Step 4 & 5: Determine the result
        std::vector<int> remaining_methods;
        if (!can_remove_suspicious) {
            // If it's not possible to remove all suspicious methods, return all methods
            for (int i = 0; i < n; ++i) {
                remaining_methods.push_back(i);
            }
        } else {
            // If all suspicious methods can be removed, return the non-suspicious ones
            for (int i = 0; i < n; ++i) {
                if (suspicious_methods.find(i) == suspicious_methods.end()) {
                    remaining_methods.push_back(i);
                }
            }
        }

        return remaining_methods;
    }
};
